### DISCUSSION TITLE
How builds under ARM64 Homebrew are supposed to be getting "-L/opt/homebrew/lib/" to LDFLAGS?

### DISCUSSION THREAD

**Discussion by native-api** [Homebrew/discussions#1423] [Everyday usage] [answered]
A user reports an ARM64 build (initiated by Pyenv) not finding some `.dylib`s installed in `/opt/homebrew/lib`: https://github.com/pyenv/pyenv/issues/1877

Looking through the diagnostic info (provided on the above link), I see that `ld` does not get `-L/opt/homebrew/lib/` into its command line -- even though the Pyenv instance does belong to the ARM64 Homebrew installation.

**I suspect that this user lacks some necessary configuration to their environment or something -- but without knowing the mechanism of where this switch should be coming from by the ARM64 installation design, I've no idea where they should look. Could someone enlighten us on this?**

(I don't have Apple hardware so I can't diagnose this myself any further.)

**Comment by carlocab** [comment, marked as answer]
Is there a way to configure `pyenv` with flags it needs to pass to the compiler by default when building Python? `/opt/homebrew/lib` will not be in the default search path for the system compiler, so it seems to me that adding `-L/opt/homebrew/lib` to `LDFLAGS` is the only fix here otherwise.

**Reply by native-api** [reply]
Of course, we can just `export` these flags. But I wonder if this is the intended way.

Since these flags have to be passed to _all_ builds done under the ARM64 Homebrew's umbrella, there must be some predefined way in it to facilitate passing them automatically!

**Reply by carlocab** [reply]
That sounds like something `pyenv` should be doing. `brew` only knows about installing and uninstalling software, and doesn't have anything to do with what a piece of software does between the two.

Of course, if there were things to configure at install time that improves usage of the software, we can do that.

How does `pyenv` handle finding libraries when a user has installed them in non-default location?

**Reply by native-api** [reply]
> `brew` only knows about installing and uninstalling software

It also knows about building software (or rather, its formulas do). Do _all_ the formulas specify these flags (`-I/opt/homebrew/include` in `CPPFLAGS` is also needed as it turns out) by hand?

_Or they expect something else to do this for them?_

**Reply by native-api** [reply]
> How does `pyenv` handle finding libraries when a user has installed them in non-default location?

By allowing to pass additional flags to the build.

But when I build a Brew formula from source in standard configuration, I don't need to pass any additional flags -- that's why I think that Homebrew itself must be providing the aforementioned two flags (and maybe others) somehow.

**Reply by SMillerDev** [reply]
> Do all the formulas specify these flags (-I/opt/homebrew/include in CPPFLAGS is also needed as it turns out) by hand?

Yes, because the brew build environment will provide these (but specific to the declared dependencies) when it builds something.

**Reply by carlocab** [reply]
> But when I build a Brew formula from source in standard configuration, I don't need to pass any additional flags -- that's why I think that Homebrew itself must be providing the aforementioned two flags (and maybe others) somehow.

`brew` adds the appropriate build flags and environment variables based on what's declared in a formula.

**Reply by native-api** [reply]
Good, that's what I was asking about!
So, you're saying that these flags are only available to formula logic by Homebrew design?

We need some way for our program (that is itself installed via Homebrew) to somehow get these for the (non-formula) builds it runs. We can provide a list of (direct) dependencies that we take from Homebrew to get a more tailored list.
If that's not possible reliably, deducing them in some more-or-less future-proof way would do.
Currently, we add `-L$(brew --prefix <dependency>)/lib` (same for `include`) (see e.g. https://github.com/pyenv/pyenv/blob/debf1abd8bd624ba8f9f902f00e5c9098e99dede/plugins/python-build/bin/python-build#L1361-L1372).

**Reply by carlocab** [reply]
You could do that. If your dependencies ship `pc` files (e.g. `readline` does) you can also use `pkg-config` to generate the appropriate build flags. `PKG_CONFIG_PATH` would need to be set appropriately for keg-only dependencies though.

**Reply by native-api** [reply]
> You could do that.

So, you're saying that the current way is the best way to go and we just need to also add `$(brew --prefix)/<lib/include>` whenever we use anything from Homebrew?

`pkg-config` doesn't seem too helpful. First, can we even rely on it to always be present? Then, its output doesn't include flags for dependencies, `$(brew --prefix)/<stuff>` (but I only have an x64 system to test on), and I already need to know the paths to find the `.pc` file.

**Reply by carlocab** [reply]
> So, you're saying that the current way is the best way to go and we just need to also add `$(brew --prefix)/<lib/include>` whenever we use anything from Homebrew?

If those are the correct library and include directories, yes. This is why I suggested relying on `pkg-config`: it's meant to be a generic, cross-platform solution for generating the correct build flags. Using `pkg-config` should work whether you're relying on dependencies from `brew`, or from some other platform or custom build that doesn't put their libraries in `lib`.

> `pkg-config` doesn't seem too helpful. First, can we even rely on it to always be present?

If you make it a dependency of `pyenv`, why not.

> Then, its output doesn't include flags for dependencies, `$(brew --prefix)/<stuff>` (but I only have an x64 system to test on)

Not sure what you mean by this. 

> I already need to know the paths to find the `.pc` file.

In general, I don't think this is the case. This probably only really applies to keg-only dependencies from `brew`, and `brew info <formula>` explains how to configure `PKG_CONFIG_PATH` correctly for these.

**Reply by native-api** [reply]
> > Then, its output doesn't include flags for dependencies, `$(brew --prefix)/<stuff>` (but I only have an x64 system to test on)
> 
> Not sure what you mean by this.

I thought that `libreadline` depends on `libintl` and expected to see flags for both in `pkg-config --libs`. It turns out that it doesn't depend on `libintl`. Checking `--libs` for _all_ `.pc` files at my system does show that flags for dependencies seem to be included.

However, I also see that Python build does locate Homebrew's `pkg-config`, and it links to `libintl` (the lib that the build can't find) directly -- but that doesn't cause `-L/opt/homebrew/lib` to be added to linker flags. The package that `libintl` is part of, `gettext`, doesn't have a `.pc` file.

So we can't rely on `pkg-config` solely, after all. :-(

**Reply by native-api** [reply]
So, adding `$(brew --prefix)/<stuff>` (and `$(brew --prefix <formula>)/<stuff>` for keg-only packages) manually is an okay and a fairly future-proof way.

---

For the record, how would I tap into the logic that provides formulas with build flags?

I guess I'd need to write a custom formula that would get this info somehow, print it and then quit immediately (so that Homebrew metadata is unchanged) -- then pass it to `brew install`?

**Reply by carlocab** [reply]
You don't see `libintl` as a dependency of `libreadline` because `libreadline` does **not** depend on `libintl`. Python does have a nasty habit of linking with `libintl` whenever it can find it, though (even if you don't want it to).

> For the record, how would I tap into the logic that provides formulas with build flags?
> 
> I guess I'd need to write a custom formula that would get this info somehow, print it and then quit immediately (so that Homebrew metadata is unchanged) -- then pass it to `brew install`?

I don't know of a reliable way to get this programmatically, I'm afraid. Part of this logic lives in `brew`'s compiler shims, which is a private API. You could try creating a formula with the right dependencies and doing `system "env"` in the install method to see the build environment variables with `brew install --verbose`. But that won't be everything.
